### DISCUSSION TITLE
Provide a mechanizm to disable SSR

### DISCUSSION THREAD

**Discussion by cherniavskii** [vercel/next.js#62352] [Ideas]
**Is your feature request related to a problem? Please describe.**
The problem is that sometimes server load is high because of SSR. Right now it can be optimized by turning off SSR for some components, but it doesn't solve the problem fully.

**Describe the solution you'd like**
Ideally - to provide something like Electrode has: [Server Side Rendering Modes](https://docs.electrode.io/chapter1/intermediate/server-side-data-hydration.html#server-side-rendering-modes).
> Electrode constantly monitors the performance of your app and turns off server-side rendering if app performance is degrading.

**Describe alternatives you've considered**
An alternative (and probably more elastic) solution would be to provide SSR Modes without any performance monitoring.

**Additional context**
I know that right now next.js is tightly coupled with SSR, but SSR modes should help a lot at large scale.

P.S. Thanks you guys for working on such a great project ;)

**Comment by mauvew** [comment]
+1

**Comment by soroushchehresa** [comment]
+2 :)))

**Comment by mzaidse** [comment]
+1

**Comment by Zertz** [comment]
While not built in, [react-no-ssr](https://github.com/kadirahq/react-no-ssr) works quite well for that purpose!

**Comment by gautamsi** [comment]
proposal: Checking for a static property on default import with { ssr: false} should make this  dynamic component without ssr. 

This will greatly help in rapid development while I figure out if component can be dealt with ssr or not.

**Comment by farhadniaz** [comment]
We should provide a flag in Nextjs config to provide Universal and SPA mode. this is very important . 
After working with Nextjs in different projects i find out Nextjs has many powerful aspects but it suffers from community support .

**Comment by timneutkens** [comment]
@farhadniaz I'm not sure what you're asking for 🤔 You can already choose to not render the tree on the server using something like react-no-ssr or a React hook.

**Comment by farhadniaz** [comment]
@timneutkens  react-no-ssr not the answer . I please look at the NUXT.js SPA mode and you will find out what i mean

**Comment by timneutkens** [comment]
I'm not going into this further as it's off-topic to this issue, this issue is related to backoff under load. Nuxt's SPA mode is quite the same as just opting out of SSR, except it's a flag, which in general you don't want, as you might want to prerender part of the skeleton.

**Comment by farhadniaz** [comment]
@timneutkens  so you mean in _app.js we have to use react-no-ssr as root component?

**Comment by nguyenvanhoang26041994** [comment]
I have the same problem and that what i resolve my.
```
// pages/_app.js
import dynamic from 'next/dynamic';

export default dynamic(() => import('./App'), { ssr: !process.env.APP_MODE === 'SPA' });

```
```
// getInitialProps function
export default async ({}) => {
  return {
    gotInitialProps: true,
  };
}
```
```
// if gotInitialProps = false, fetch API
useSPAMode(gotInitialProps, getInitialProps);
```
```
import React, { useEffect } from 'react';
import useStore from './useStore';

const useSPAMode = (gotInitialProps, getInitialProps) => {
// getInitialProps maybe using store. i using redux for nextjs
  const { dispatch, getState } = useStore();
  useEffect(() => {
    if (!gotInitialProps) {
      getInitialProps({
        reduxStore: {
          dispatch,
          getState,
        },
        isServer: false,
      });
    }
  }, []);
};

export default useSPAMode;
```

**Comment by sophister** [comment]
Does Next.js provide such mechanizm now? I can't find any document related to this.

**Comment by benalexb** [comment]
Bumping this issue up, as it has become quite common these days to use Next as a static site generator. Of course, one can always build, export, and serve the static website, but that comes at a cost (live-reloading and other development sugar).

There are other motivations too behind the ability to disabled SSR, such as to quash react's warning when the server-side rendered DOM does not match the client's, which is very common on a static website as there is no pre-loaded data.

**Comment by Hubro** [comment]
I want to be able to temporarily disable server side rendering so I can debug a crash in my browser with dev tools rather than trying to figure out from the dev server log what happened. Installing a new dependency and wrapping my code in a new element seems like an excessive requirement to achieve this :/

I just want something like `next --no-ssr` until I'm done debugging the error, then I remove the flag again. Why would something like that be problematic to implement? Couldn't you even just include `react-no-ssr` by default and wrap the topmost element in `<NoSSR>` if the flag is provided?

**Comment by lostfictions** [comment]
`react-no-ssr` won't help if you have side-effectful imports that rely on the DOM. But you can use next's built-in dynamic import functionality instead, and move your page root (or whatever subset of it can't run on the server) into a different file:

```jsx
import dynamic from 'next/dynamic'
const App = dynamic(() => import('../components/App'), { ssr: false })
export default () => <App />
```

This works well enough for me as someone making a static site that relies on eg. WebGL and DOM APIs, but who still wants to take advantage of Next's sensible defaults and configurability (rather than, say, trying to puzzle out which variant of `create-react-app` override packages is still maintained).

---

This maybe also suggests a sketch for folks who want to toggle SSR with a switch on the command line or disable it in development and enable it during build for static export:

```jsx
import dynamic from 'next/dynamic'
const App = dynamic(() => import('../components/App'), {
  ssr: process.env.NEXT_PUBLIC_SSR_ENABLED
})
export default () => <App />
```

Then, on the command-line:
```
NEXT_PUBLIC_SSR_ENABLED=false npm run dev
```

That said, I personally think it would still be great if Next added better support for non-SSR scenarios -- as mentioned in this thread, there's still lots of idiosyncrasies for people using Next for fully static (non-hybrid) static site generation, etc.

**Comment by renatodex** [comment]
Dynamic imports really make our life easier, but if you think about Canvas integration, there are a few issues using the SSR approach. I've tried to use `PhaserJS` before, which relies on Dom at the very moment it is imported, and I never could make it load as a component.

Of course, Canvas frameworks like Phaser don't really play well with React and Virtual Dom, but even a loose abstraction is hard to assemble with SSR.

**Comment by timneutkens** [comment]
@renatodex you can dynamically import files on client-side using `import()`

**Comment by renatodex** [comment]
Hey @timneutkens !
I don't want to distort the main goal of this Issue, so I've posted my failed experiment with `Phaser` as a new Issue in the Issues of NextJS.
https://github.com/zeit/next.js/issues/13417
Take a look when you have time!

**Comment by eric-burel** [comment]
Not sure if it has been mentioned before, but easy and global disable of SSR is useful in the context of e2e/integration testing.

Cypress provides mock for requests triggered by the browser, but it's much more difficult to mock server side queries, especially when they hit the same origin. There's not much recipe to handle this on the web, so the easiest path is to disable SSR when running e2e tests.

**Comment by jgoux** [comment]
> `react-no-ssr` won't help if you have side-effectful imports that rely on the DOM. But you can use next's built-in dynamic import functionality instead, and move your page root (or whatever subset of it can't run on the server) into a different file:
> 
> ```js-jsx
> import dynamic from 'next/dynamic'
> const App = dynamic(() => import('../components/App'), { ssr: false })
> export default () => <App />
> ```
> 
> This works well enough for me as someone making a static site that relies on eg. WebGL and DOM APIs, but who still wants to take advantage of Next's sensible defaults and configurability (rather than, say, trying to puzzle out which variant of `create-react-app` override packages is still maintained).
> 
> This maybe also suggests a sketch for folks who want to toggle SSR with a switch on the command line or disable it in development and enable it during build for static export:
> 
> ```js-jsx
> import dynamic from 'next/dynamic'
> const App = dynamic(() => import('../components/App'), {
>   ssr: process.env.NEXT_PUBLIC_SSR_ENABLED
> })
> export default () => <App />
> ```
> 
> Then, on the command-line:
> 
> ```
> NEXT_PUBLIC_SSR_ENABLED=false npm run dev
> ```
> 
> That said, I personally think it would still be great if Next added better support for non-SSR scenarios -- as mentioned in this thread, there's still lots of idiosyncrasies for people using Next for fully static (non-hybrid) static site generation, etc.

@lostfictions @timneutkens Does this strategy allows to use something like React-router and being able to export the app as a single index.html + multiple js files (code-splitted)?
I'm in the same boat as you @lostfictions , I want to take advantage of Next.js extendable configuration in place of CRA, but I don't want to have a server coupled with my SPA.
I'm aware of `next export` but I need a single html entry point to my app (in order to wrap it with capacitor for mobile distribution).

**Comment by verekia** [comment]
It would be very convenient for debugging purposes to have this `mode` option, similar to what Nuxt has. But to go even further, it would also be convenient to have an `ssr-only` mode. This is very useful to debug universal authentication for instance. You can make sure everything works during client navigation only, and during server navigation only. I used to do this when I made my own isomorphic apps without Next and I really miss being able to do that.

- `next dev --mode=spa-only`
- `next dev --mode=ssr-only`
- `next dev --mode=universal` (default)

Something like that.

It was not that complicated when I was doing it by hand. For SPA-only I simply rendered an empty main app root element for every page, with no server data, and for SSR-only, I didn't include any JS file in the HTML. Now of course it's a bit more complicated in the Next architecture, but both are quite simple in theory.

Edit: I just realized that for development purposes, SSR-only can be achieved by just disabling JavaScript in the browser.

**Comment by eric-burel** [comment]
Another use case I have for SSR disabling is education. 
It would be very cool to be able to first learn Next without SSR, then activate SSR and show what it breaks, and then learn how to handle client-only code gracefully.

I strongly felt this need while trying to teach Next to data science students that are used to heavy client-only libraries like Chartjs, D3, leaflet etc. In the future I would probably favour Create React App for this reason, but I'd prefer keeping Next.

**Comment by johot** [comment]
I have created a web app that I host inside a React Native WebView. I love the developer experience of Next.js but in this case I have no use for SSR it just makes thing more complicated, gives me layout issues for non SSR supported hooks etc etc. In this case I want to use react router for the routing because of the support for nested routes that don't loose state and so on.

I would love an easy and official way of disabling SSR, right now I am following this guide by @tannerlinsley which is great:
https://gist.github.com/tannerlinsley/65ac1f0175d79d19762cf06650707830

I definately see a lot of use cases for a "SPA mode / no SSR mode".

Thank you for an awesome product @timneutkens !

**Comment by amirhbakan** [comment]
next export + next/router and a nginx that points for example / can work like SPA but it’s not the best solution, because next export generates many htmls and also if we accept multi htmls there is no way to point pages like product/[id] with nginx but it handled clientside with next/router.
I think next export + next/router could be a good solution for SPA mode with some features,
1- really generate one html or a solution for pointing dynamic routes pages with nginx.
2- handle 404 clientside in next/router with a flag.

**Comment by eric-burel** [comment]
> next export + next/router and a nginx that points for example / can work like SPA but it’s not the best solution, because next export generates many htmls and also if we accept multi htmls there is no way to point pages like product/[id] with nginx but it handled clientside with next/router. I think next export + next/router could be a good solution for SPA mode with some features, 1- really generate one html or a solution for pointing dynamic routes pages with nginx. 2- handle 404 clientside in next/router with a flag.

That's the point of the proxy here. Last time I tested (a while ago), Next would produce a `[id].html` in fallback mode. You only get many HTML, eg one per post of your blog, if you don't use fallback. Your Nginx can point to this `[id].html` which would then be able to do client-side fetching to get the right content.
This is faster than an SPA because your code is split per page.

I agree that a pure SPA mode could be interesting as well, as it is the simplest way to export. However I've kinda dropped this idea a while ago, because I prefer the way Next uses the presence of the server intelligently.
If you really need an SPA, then Create React App might be a simpler alternative.

**Comment by amirhbakan** [comment]
> > next export + next/router and a nginx that points for example / can work like SPA but it’s not the best solution, because next export generates many htmls and also if we accept multi htmls there is no way to point pages like product/[id] with nginx but it handled clientside with next/router. I think next export + next/router could be a good solution for SPA mode with some features, 1- really generate one html or a solution for pointing dynamic routes pages with nginx. 2- handle 404 clientside in next/router with a flag.
> 
> That's the point of the proxy here. Last time I tested (a while ago), Next would produce a `[id].html` in fallback mode. You only get many HTML, eg one per post of your blog, if you don't use fallback. Your Nginx can point to this `[id].html` which would then be able to do client-side fetching to get the right content. This is faster than an SPA because your code is split per page.
> 
> I agree that a pure SPA mode could be interesting as well, as it is the simplest way to export. However I've kinda dropped this idea a while ago, because I prefer the way Next uses the presence of the server intelligently. If you really need an SPA, then Create React App might be a simpler alternative.

Thanks for the reply, but I need this because I have implemented the Dynamic rendering with NextJS, I want to have one codebase with an SSR build and a CSR build, my SSR side is just SSR, not SSG. we implemented it with some tricks, but I want a good solution for CSR build to really have a SPA. I think if these features implement, the Dynamic Rendering with NextJS is a robust technique for websites.

**Comment by habovh** [comment]
For anyone looking for a simple and zero additional dependency solution, I suggest wrapping your app in a component that you can import using Next `dynamic()` import. Then you can check the router for a set query parameter, allowing you to easily debug issues on a one-off basis.

Something like this would *"disable"* SSR for any component below the `<Main />` node when the URL contains the query string `?disable-ssr=1`.

```jsx
// _app.js
import dynamic from 'next/dynamic'

import MainSSR from '../components/main'

const MainNoSSR = dynamic(() => import('../components/main'), { ssr: false })

class MyApp extends App {
  render() {
    const { Component, router } = this.props
    const disableSSR = router.query['disable-ssr'] === '1'
    const Main = disableSSR ? MainNoSSR : MainSSR
    return (
      <>
        <Head>
          {...}
        </Head>
        <Main>
          <Component />
        </Main>
      </>
    )
  }
}
```

**Comment by verekia** [comment]
@habovh This is really great. What's in `MainSSR`? Just something like `const Main = ({ children }) => <>{children}</>`?

**Comment by habovh** [comment]
@verekia yup exactly. It's also handy to add content that must be present on every page —such as the cookie banner alongside the children.

Edit: just to be precise, I'm referring to the component exported in `components/main` that is being imported by both `MainSSR` and `MainNoSSR` using two different import methods.

**Comment by paulogaspar7** [comment]
New version of Ant Design's Pro-Components also not supporting SSR.
(Node complains with a "SyntaxError: Cannot use import statement outside a module".)

**Comment by remorses** [comment]
I just created a [Next.js plugin](https://github.com/remorses/elacca) to disable SSR and remove all the React components from your server bundle.

This plugin will transform your pages to only return `null` in the server, causing dead code elimination to remove all the React components and libraries from your server output code. On the client, the pages will return `null` until mounted.

To use it simply add the `skip ssr` directive at the top of your pages:

```js
// pages/index.js
'skip ssr'

export default function Home() {
    return <div>hello world</div>
}
```

Read more in the [Elacca repository](https://github.com/remorses/elacca)

**Comment by samcx** [comment]
Hi everyone!

Since this is a feature request, I will be moving this over to <picture data-single-emoji=":nextjs:" title=":nextjs:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/nextjs/2a85d633ae594556.png" alt=":nextjs:" align="absmiddle"></picture> discussions. <picture data-single-emoji=":loading:" title=":loading:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/loading/fa531af0d6940856.gif" alt=":loading:" align="absmiddle"></picture>

**Comment by ethanjdiamond** [comment]
Hi Nextjs team!

I'm currently running into the need to disable SSR, but only at the single component level. This is not super common, but comes up fairly often when you're rendering something like "X seconds ago" where the client and server will pretty much always disagree and cause a hydration error. While I see in the [documentation](https://nextjs.org/docs/app/building-your-application/optimizing/lazy-loading) that you can use `const ComponentC = dynamic(() => import('../components/C'), { ssr: false }`, which will work for my use case, I think there's room for improvement here.

I think a lot of the confusion that comes with next.js comes from there only being "Server Component" and "Client Component" when it's really describing a table with three slots:

|                                 | Server Render |  No Server Render  |
|------------------ | -------------- | ---------------------- |
| **Client Render**       |  Client Component     | [This spot has no component "type"] |
| **No Client Render** |  Server Component    | *N/A*  |

A pretty good solution here is to work with the React team to add a Client-Only Component type as a first class citizen and a "use client-only" directive. Server components can have Server, Client and Client-Only Components as children, Client Components can have Client and Client-Only Components as children, and Client-Only Components can only have Client-Only Components as children. It would work for my use case of having a component where server and client will usually disagree, but would also work for the use case of the above teams who want full no-SSR as they can just make their root "use client-only". Win-win.

As an aside, I think all of the confusion around rendering would be gone if this was the naming convention, but that ship has sailed:

|                                | Server Render |  No Server Render  |
|------------------ | -------------- | ---------------------- |
| **Client Render**       | Mixed Component     | Client Component |
| **No Client Render** | Server Component    | *N/A*  |

**Reply by zacaj** [reply]
This makes so much sense, it makes it hard to believe that deep in a random half-unrelated feature request is the only place I've seen it discussed.

Having to use dynamic imports at EVERY place that uses a component doesn't make sense for client-only components, it should be something that's marked within that component directly.  "use client-only" would solve this perfectly

**Comment by ocano-diego** [comment]
For my project I don't want to EVER use SSR, I will SSG all my pages, forever.

**Comment by verekia** [comment]
Just wanted to clarify something, as SSR is an ambiguous term. With SSG, there is still React's SSR which is run at build-time to create the initial HTML, which is a different thing from Next.js' dynamic SSR. Some of us are talking about disabling React's SSR to have the HTML empty at page load and behave like a pure SPA (which can help with debugging), while others are talking about disabling Next/Vercel's dynamic SSR and serverless functions.

Disabling React's SSR can be done by wrapping the app/pages in a ClientOnly component, as described in this post: https://www.joshwcomeau.com/react/the-perils-of-rehydration/

**Reply by AntonioRedondo** [reply]
Which is sad that out-of-the-box Next.js does not allow to tweak or disable SSR, and framework consumers have to resort to a custom component to get this basic behaviour.

**Comment by miracoco** [comment]
+1, some components, such as OpenLayers, that completely depend on the browser environment. When integrating those thrid-party components need to handle SSR carefully.

**Comment by alicercedigital** [comment]
This feature is a must. With it, we address the use case that previously required us to use Vite just to get good old client-side React.
